src/plonkish/backend/halo2.rs

```diff
@@ -514,3 +679,153 @@ impl<F: PrimeField + From<u64> + Hash> h2Circuit<F> for ChiquitoHalo2SuperCircui
+    for phase in 0..cs.phases() {
+        println!("phase {phase}");
+
+        let mut assigned_witness =
+            vec![Some(vec![Fr::default(); params.n() as usize]); cs.num_advice_columns];
+
+        // TODO as super_circuit.get_mapping().generate() creates a mapping of "compilation unit" ID
+        // to a witness, currently it is impossible to get a witness by the circuit id, hence we are
+        // iterating over all witnesses and checking if the circuit has a column in
+        // question.
+        for witness in witnesses.iter() {
+            for (column, assignments) in witness.iter() {
+                for circuit in circuits.iter() {
+                    if let Some(circuit_column) = circuit.advice_columns.get(&column.uuid()) {
+                        let halo2_column = Column::<Any>::from(*circuit_column);
+                        for (offset, value) in assignments.iter().enumerate() {
+                            assigned_witness[halo2_column.index].as_mut().unwrap()[offset] = *value;
+                        }
+                    }
+                }
+            }
+        }
+
+        // TODO ignoring the challenges produced by the phase, but can they be useful later?
+        let _ = prover.commit_phase(phase as u8, assigned_witness).unwrap();
+        println!("phase {phase} completed");
+    }
+    prover.create_proof().unwrap();
+    let proof = transcript.finalize();
+    println!("Prove: {:?}", start.elapsed());
+
+    proof
+}
+
+#[allow(clippy::type_complexity)]
+pub fn get_halo2_setup(
+    k: u32,
+    circuit: Circuit<Fr>,
+    rng: BlockRng<OneNg>,
+) -> (
+    ConstraintSystemMid<Fr>,
+    ParamsKZG<Bn256>,
+    VerifyingKey<G1Affine>,
+    ProvingKey<G1Affine>,
+    ChiquitoHalo2<Fr>,
+) {
+    let mut chiquito_halo2 = ChiquitoHalo2::new(circuit);
+
+    let compiled = compile_halo2_middleware::<Fr>(k, &mut chiquito_halo2).unwrap();
+    // Setup
+    let params = ParamsKZG::<Bn256>::setup::<BlockRng<OneNg>>(k, rng);
+    let vk = keygen_vk(&params, &compiled).expect("keygen_vk should not fail");
+    let pk = keygen_pk(&params, vk.clone(), &compiled).expect("keygen_pk should not fail");
+
+    (compiled.cs, params, vk, pk, chiquito_halo2)
+}
+
+pub fn get_super_circuit_halo2_setup(
+    k: u32,
+    circuit: &mut ChiquitoHalo2SuperCircuit<Fr>,
+    rng: BlockRng<OneNg>,
+) -> (
+    ConstraintSystemMid<Fr>,
+    ParamsKZG<Bn256>,
+    halo2_proofs::plonk::VerifyingKey<G1Affine>,
+    halo2_proofs::plonk::ProvingKey<G1Affine>,
+) {
+    let compiled = compile_super_circuit_halo2_middleware::<Fr>(k, circuit).unwrap();
+    // Setup
+    let params = ParamsKZG::<Bn256>::setup::<BlockRng<OneNg>>(k, rng);
+    let vk = keygen_vk(&params, &compiled).expect("keygen_vk should not fail");
+    let pk = keygen_pk(&params, vk.clone(), &compiled).expect("keygen_pk should not fail");
+
+    (compiled.cs, params, vk, pk)
+}
+
+// One number generator, that can be used as a deterministic Rng, outputting fixed values.
```

Comment:
@alxkzmn great idea, but the name for me is confusing, why `OneNg`? Perhaps `MockRng`? Something that denotes that should not be used in production.

Reply:
I copied it directly from the Halo2 repo 😄 I'll change the name